python/tank/platform/engine.py

```diff
@@ -959,33 +1001,59 @@ def __create_main_thread_invoker(self):
         """
         Create the object used to invoke function calls on the main thread when
         called from a different thread.
-        
+
         :returns:  Invoker instance
         """
         if self.has_ui:
             from .qt import QtGui, QtCore
             if QtGui and QtCore:
                 class Invoker(QtCore.QObject):
+                    """
+                    Invoker class - implements a mechanism to execute a function with arbitrary
+                    args in the main thread.
+                    """
                     def __init__(self):
+                        """
+                        Construction
+                        """
                         QtCore.QObject.__init__(self)
+                        self._lock = threading.Lock()
+                        self._fn = None
                         self._res = None
-                        
+
                     def invoke(self, fn, *args, **kwargs):
-                        self._fn = lambda: fn(*args, **kwargs) 
-                        self._res = None
-                        
-                        QtCore.QMetaObject.invokeMethod(self, "_do_invoke", QtCore.Qt.BlockingQueuedConnection)
-                        
-                        return self._res
-                
+                        """
+                        Invoke the specified function with the specified args in the main thread
+
+                        :param fn:          The function to execute in the main thread
+                        :param *args:       Args for the function
+                        :param **kwargs:    Named arguments for the function
+                        :returns:           The result returned by the function
+                        """
+                        # acquire lock to ensure that the function and result are not overwritten 
+                        # by syncrounous calls to this method from different threads
+                        self._lock.acquire()
```

Comment:
Creating the invoker when first needed would require a look to see if it has been created, so now you'd be locking twice, so creating it upfront is the better solution.

I think this whole scheme of mine was a bit pointless tough, since even if you could launch the calls in parallel, they are still going to be processed by the main thread in a FIFO fashion, so there's no benifit really to a lockless approach.

Regardless, your point about performance vs stability is sound. stability > performance.